Open Liberty/Microprofile running on IBM Code Engine with CloudEvents (#2636)
OpenLiberty/blogs: posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

---
layout: post
title: "MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/abutch3r
author_github: https://github.com/abutch3r
seo-title: MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine - OpenLiberty.io
seo-description: How to run MicroProfile applications in the IBM Cloud Code Engine serverless environments using CloudEvents
blog_description: "Running MicroProfile applications in the IBM Cloud Code Engine serverless functions"
additional_authors:
- name: Emily Jiang
  github: https://github.com/Emily-Jiang
  image: https://avatars0.githubusercontent.com/Emily-Jiang
---

> = MicroProfile applications running as a serverless function via CloudEvent in IBM Cloud Code Engine

Reviewer: ```suggestion
= Run MicroProfile applications as a serverless function with CloudEvent in IBM Cloud Code Engine
```

Alex Butcher <https://github.com/abutch3r>
:imagesdir: /
:url-prefix:
:url-about: /

> Serverless is certainly a hot topic in 2022. MicroProfile has been adopted widely by cloud native applications. An often question being asked is:

Reviewer: ```suggestion
Serverless applications, which can scale to zero and allocate resources on demand, are certainly a hot topic in 2022. With MicroProfile widely adopted by cloud native applications, many developers want to know:
```

> Can MicroProfile Applications run as a serverless?

Reviewer: ```suggestion
_Can MicroProfile applications be serverless?_
```

> The short answer is YES. This blog explains the basic concept of Serverless, CloudEvent and IBM Cloud Code Engine and then demonstrate how to get an MicroProfile application running on Open Liberty running as a serverless function in IBM Cloud Code Engine via CloudEvent.

Reviewer: ```suggestion
The short answer is YES. This post demonstrates how to get a MicroProfile application that runs on Open Liberty to run as a serverless function in IBM Cloud Code Engine by using CloudEvents.
```

> *Let's first explain what is Serverless!*

Reviewer: ```suggestion
```

> == Serverless

Reviewer: ```suggestion
== What is serverless?
```

> Serverless is a short name for Serverless Computing, which is an execution model in which the cloud provider allocates resources on demand. Serverless enables developers to concentrate on their own applications without the need of managing servers. It does not mean there is no server running in the background. On a contrary, Serverless architecture contains servers.

Reviewer: ```suggestion
Serverless is short for serverless computing, which is an execution model in which the cloud provider allocates resources on demand. Serverless enables developers to concentrate on their own applications without needing to manage servers. It does not mean no server is running in the background. On the contrary, serverless architecture contains servers.
```

> Serverless normally executes code in the form of a function, so Serverless sometimes referred to “Functions as a Service” (FaaS) or "Container as a Service" (CaaS). There are a few from major cloud providers supporting serverless:

Reviewer: ```suggestion
Serverless normally executes code in the form of a function and is sometimes referred to “Functions as a Service” (FaaS) or "Container as a Service" (CaaS). There are a few major cloud providers that support serverless:
```

*	IBM Cloud Code Engine: IBM Cloud Function
*	AWS: AWS Lambda
*	Microsoft Azure: Azure Functions
*	Google Cloud: Google Cloud Functions

> Serverless is a model. In order to work with Serverless, Knative and Cloud Event are often used.  https://knative.dev/docs/concepts/[Knative] provides a platform-agnostic solution for running serverless deployment. CloudEvent is also a common event model to be used in Serverless architecture. This blog will focus on CloudEvent instead.

Reviewer: ```suggestion
```

== CloudEvents

> https://cloudevents.io/[CloudEvents] is a specification of describing event data in a common and language agnostic way. CloudEvents is to simplify event declaration and delivery across the services and platforms. This specification is under Cloud Native Computing Foundation (CNCF). CloudEvents are supported by many Serverless frameworks including IBM Cloud Code Engine, which will be discussed in the next section.

Reviewer: ```suggestion
https://knative.dev/docs/concepts[Knative] and https://cloudevents.io/[CloudEvents] are  two frameworks that are often used to work with the serverless model. This post focuses on CloudEvents. 

CloudEvents is a specification to describe event data in a common language-agnostic way. CloudEvents simplifies event declaration and delivery across the services and platforms. This specification is under the Cloud Native Computing Foundation (CNCF). CloudEvents are supported by many serverless frameworks, including IBM Cloud Code Engine.
```

== IBM Cloud Code Engine

> https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine](ICCE) is a fully managed, serverless platform designed to run container images, batch jobs, or source code. Code Engine allows you to build event driven workloads that react to CloudEvents. Code Engine is truly serverless. It automatically scales your workloads up and down, and even down to zero when there are no requests. You only pay for the resources you consume. You can easily run your cloud native applications in Code Engine. The common cloud native applications normal use MicroProfile specifications for configuring, securing, and observing applications.

Reviewer: ```suggestion
https://www.ibm.com/cloud/code-engine[IBM Cloud Code Engine](ICCE) is a fully managed, serverless platform designed to run container images, batch jobs, or source code. Code Engine allows you to build event driven workloads that react to CloudEvents. Code Engine is truly serverless. It automatically scales your workloads up and down, even down to zero when there are no requests. You pay for only the resources you consume. You can easily run your cloud native applications in Code Engine. 
```

== MicroProfile Applications

> MicroProfile Applications arhe popular cloud native applications. MicroProfile offers a set of standards APIs for cloud-native applications that free your applications from vendor lock-in. Open Liberty is the leading implementation for MicroProfile specifications and we will use Open Liberty in this blog.

Reviewer: ```suggestion
Many developers use MicroProfile specifications for configuring, securing, and observing cloud native applications. MicroProfile offers a set of standards APIs for cloud-native applications that free your them from vendor lock-in. Open Liberty is the leading implementation for MicroProfile specifications.
```

> Since we are entering into Serverless era, it will be very important to get MicroProfile Applications running in a serverless environment. Luckily with CloudEvent, it is very straightforward to get these kinds of applications running as functions in an environment that supports CloudEvents such as IBM Cloud Code Engine. In the next section, we will demonstrate how to get MicroProfile application deployed to Open Liberty running as a function in IBM Cloud Code Engine via CloudEvent.

Reviewer: ```suggestion
As we enter the serverless era, it is very important to get MicroProfile applications running in a serverless environment. Luckily, it is very straightforward to get these applications running as functions in an environment that supports CloudEvents, such as IBM Cloud Code Engine. In the following sections, we demonstrate how to get a MicroProfile application that is deployed to Open Liberty running as a function in IBM Cloud Code Engine with CloudEvent.
```

> == MicroProfile applications running as a serverless in Code Engine

Reviewer: ```suggestion
== Run MicroProfile applications  as a serverless function in Code Engine
```

> RESTful MicroProfile Applications can run seamlessly as a serverless in Code Engine via CloudEvent. This section will explain in a step-by-step manner.

Reviewer: ```suggestion
RESTful MicroProfile Applications can run seamlessly as a serverless function in Code Engine with CloudEvent. This section explains how.
```

> === Pre-requisites

Reviewer: ```suggestion
=== Prerequisites
```

> Before you start you will need the following

Reviewer: ```suggestion
Before you start you will need the following prerequisites:
```

> * IBM Cloud Account

Reviewer: ```suggestion
* An IBM Cloud Account
```

* Docker

> * Container Registry that is externally accessible which can host your images (IBM Cloud Container Registry is available as part of your cloud account)

Reviewer: ```suggestion
* An externally accessible container Registry that can host your images (IBM Cloud Container Registry is available as part of your cloud account)
```

> === Creating your MicroProfile Application

Reviewer: ```suggestion
=== Creating your MicroProfile application
```

> If you don't have a suitable RESTful application to host as a set of functions you can use https://openliberty.io/guides/getting-started.html[OpenLiberty and MicroProfile Getting Started] to get a simple application well suited to a Serverless environment such as IBM Cloud Code Engine.

Reviewer: ```suggestion
If you don't have a suitable RESTful application to host as a set of functions you can use the https://openliberty.io/guides/getting-started.html[OpenLiberty Getting Started guide] to get a simple MicroProfile application that is  well suited to a serverless environment, such as IBM Cloud Code Engine.
```

> It is recommended that if using the Getting Started code and you have experience with OpenLiberty and MicroProfile that you use the `finish` folder as your starting point and go to the next step. If you do not, follow the guide up to the Running the application in a Docker container step.

Reviewer: ```suggestion
If using the Getting Started code and you already have experience with OpenLiberty and MicroProfile, you can use the `finish` folder as your starting point and go to the next step. If you do not, complete the guide up to and including the https://openliberty.io/guides/getting-started.html#running-the-application-in-a-docker-container[Running the application in a Docker container] step to generate the application you'll use in this post..
```

> The step that needs to be completed before progressing is https://openliberty.io/guides/getting-started.html#running-the-application-in-a-docker-container[Running the application in a Docker Container].

Reviewer: ```suggestion
```

> At this step it is recommended that your test the application at this stage to ensure that it is running and accessible.

Reviewer: ```suggestion
After you have the application, the next step is to  test it to ensure that it is running and accessible.
```

> To run your container locally on its `http` port:

Reviewer: ```suggestion
Run the following command to run your container locally on its `http` port:
```

```
docker run -p 9080:9080 openliberty-getting-started:1.0-SNAPSHOT
```

> To invoke the application you can either use your browser, curl or an API testing tool. For example you can use the following curl command:

Reviewer: ```suggestion
To invoke the application, you can use either your browser, curl, or an API testing tool. For example you can use the following curl command:
```

```
curl http://localhost:9080/dev/system/properties
```

> This invokes the `getProperties()` method in the SystemResource class

Reviewer: ```suggestion
This command invokes the `getProperties()` method in the application's `SystemResource` class
```

> === Deploying your application to a container registry ready for IBM Cloud Code Engine

Reviewer: ```suggestion
=== Deploying your application to a container registry to prepare for IBM Cloud Code Engine
```

> Having sucessfully built and tested your image local, it now needs to be put somewhere that IBM Cloud Code Engine can get to it.

Reviewer: ```suggestion
Having sucessfully built and tested your image locally, you now need to put it somewhere that IBM Cloud Code Engine can get to it.
```

ICCE supports two primary methods to get your code into a runnable format

> 

Reviewer: I would rework this section to focus on ICCR and just mention the other alternatives at the end, keep the user working through the task in the shortest/easiest path. As it is, it's hard to tell you need to follow the ICCR link to continue. Something like

Since you already have an IBM Cloud Account, IBM Cloud Container Registry (ICCR) provides the quickest route to getting your image up and running. Follow the instructions to https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[deploy your application to ICCR].

Alternatively, ICCE supports getting your code into a running format from https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code[a source code respository] e.g. Github, a https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app[Public Registry] e.g. DockerHub, or a https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-private[Private Registry]. FOr the purposes of this post, we'll use ICCR.

* https://cloud.ibm.com/docs/codeengine?topic=codeengine-app-source-code[From source code respository] e.g. Github
* From a externally accessible Container Registry:
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app[Public Registry] e.g. DockerHub
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[IBM Cloud Container Registry] - available as part of your IBM Cloud account
** https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-private[Private Registry]

Given to follow this blog, you will have an IBM Cloud Account, IBM Cloud Container Registry (ICCR) provides the quickest route to getting your image up and running.

> === Creating your IBM Cloud Code Engine Application

Reviewer: ```suggestion
=== Creating your IBM Cloud Code Engine application
```

> With your image uploaded, you now need to create and configure you IBM Cloud Code Engine Application to make your function available.

Reviewer: ```suggestion
With your image uploaded, you now need to create and configure you IBM Cloud Code Engine application to make your function available.
```

If you used ICCR follow these https://cloud.ibm.com/docs/codeengine?topic=codeengine-deploy-app-crimage[steps to create your application using ICCR]. If your used one of the other types of registry follow the matching Application creating instructions.

> When creating your application there are some considerations to make:

Reviewer: ```suggestion
When you create your application, consider the following options:
```

> * Image reference. While the name of your image will stay the same, you will be updating it later so you should consider whether to use the image hash that is within the registry or if you want to use the tag you uploaded with that, just be aware that an update to the image may not be reflected when executed in ICCE if it uses the old tag.

Reviewer: ```suggestion
Image reference::
While the name of your image will stay the same, you will be updating it later.  Consider whether to use the image hash that is within the registry or the tag you uploaded with that.  Be aware that an update to the image may not be reflected when executed in ICCE if it uses the old tag.
```

I'm not sure what "that" means in "the tag you uploaded with that.." but this part seems confusing and I'm not sure what to suggest because I don't understand the context.

> * Resource allocation. As part of the application definition, you can tell ICCE how much CPU and Memory to allocate to any running instances. While the application is small, it is still a Java application that needs a period of startup before it can start serving requests. The Getting Started image will eventually start on the minimum values, but giving it slightly more will significantly improve startup and response time.

Reviewer: ```suggestion
Resource allocation:: 
As part of the application definition, you can tell ICCE how much CPU and Memory to allocate to any running instances. While the application is small, it is still a Java application that needs a period of startup before it can start serving requests. The Getting Started image will eventually start on the minimum values, but giving it slightly more will significantly improve startup and response time.
```

> * Listening port. Port 9080 should be used as the value for the listening port. Further information about HTTP handling within ICCE can be found https://cloud.ibm.com/docs/codeengine?topic=codeengine-application-workloads#considerationshttphandlingapp[here]

Reviewer: ```suggestion
* Listening port::
Use port 9080 as the value for the listening port. For more information, see https://cloud.ibm.com/docs/codeengine?topic=codeengine-application-workloads#considerationshttphandlingapp[Considerations for HTTP handling].
```

> Having created your application ensure that the application is not showing any errors such as `Missing Pull credentials` indicating that the image cannot be pulled to run. If you have any of these errors follow the steps in the ICCE documentation to resolve. Some errors may only occur when the application is invoked for the first time.

Reviewer: ```suggestion
After you create your application, ensure that it is not showing any errors such as `Missing Pull credentials`, which indicates that the image cannot be pulled to run. If you have any of these errors, follow the steps in the ICCE documentation to resolve. Some errors may occur only when the application is invoked for the first time.
```

=== Invoking your application on IBM Cloud Code Engine

> Having created the application, you can now invoke it within ICCE

Reviewer: ```suggestion
Now that you have the application, you can invoke it within ICCE
```

> As part of creating your app you will have been able to obtain the URL from the test application or the CLI. if you did not obtain it via these means you can follow https://cloud.ibm.com/docs/codeengine?topic=codeengine-access-servICCE[accessing your app].

Reviewer: ```suggestion
As part of creating your application, you obtained the URL from the test application or the CLI. if you did not obtain it via these means, you can follow https://cloud.ibm.com/docs/codeengine?topic=codeengine-access-servICCE[accessing your app].
```

If you make a request to `https://{ICCE_Application_URL}/` you should see the `Welcome to Open Liberty` page

> To call the application we use the same path as used for the application locally

Reviewer: ```suggestion
To call the application on ICCE, we use the same path that we for the application locally. Run the following `curl` command
```

```
curl https://${ICCE_Application_URL}/dev/system/properties
```

> After a short time you should see a response with a JSON payload containing all the system properties.

Reviewer: ```suggestion
After a short time, you should see a response with a JSON payload containing all the system properties.
```

Note: All ICCE connections are HTTPS. so while `http` was used locally, the image is configured to support HTTPS without any changes.

> Congratulations you have now called your application on IBM Cloud Code Engine.

Reviewer: ```suggestion
Congratulations! You just called your application on IBM Cloud Code Engine.
```

> While you are making the requests if you have the IBM Cloud Code Engine UI open on your applications Overview tab you will be able to see the active instances. you will need to be on the page when you start making the requests it creating the instance and then scaling down to zero when it stops receiving work

Reviewer: ```suggestion
 If you open the IBM Cloud Code Engine UI to your application's **Overview** tab, you can see the active instances while you are making requests. You will use this page when you start making requests to observe IBM Cloud Code Engine creating the application instance and then scaling down to zero when it stops receiving work.
```

image:/img/blog/ibm-code-engine-active-instances.png[IBM Cloud Code Engine Active Instances,width=70%,align="center"]

> A common use case of serverless applications is to process events coming from non-http sources such as Kafka topics or Object Storage. Historically to consume these events the application had to make use of the Cloud Providers SDK, locking the application into that provider. This is where CloudEvents comes in, as it provides a cross-provider standard around which data can be transmitted and received by an application improving portability and reducing large dependencies.

Reviewer: ```suggestion
A common use case for serverless applications is to process events coming from non-HTTP sources, such as Kafka topics or IBM Cloud Object Storage. Historically, to consume these events the application had to make use of the cloud provider's SDK, which locks the application into that provider. This is where CloudEvents comes in, as it provides a cross-provider standard around which data can be transmitted and received by an application, improving portability and reducing large dependencies.
```

> In this next section, we will demonstrate how to update your application to use CloudEvents.

Reviewer: ```suggestion
```

=== Updating the application to use CloudEvents

> IBM Cloud Code Engine connects IBM Cloud event providers such as https://cloud.ibm.com/docs/codeengine?topic=codeengine-working-kafkaevent-producer#setup-kafka-receiverapp[Kafka] or https://cloud.ibm.com/docs/codeengine?topic=codeengine-eventing-cosevent-producer#obstorage_ev_app[Object Storage] and the application via the use of CloudEvents. These events are sent to a subscribed application as HTTP POST requests. As such a RESTful application can receive these events without the requirement of additional libraries and configuration.

Reviewer: ```suggestion
IBM Cloud Code Engine connects IBM Cloud event providers such as https://cloud.ibm.com/docs/codeengine?topic=codeengine-working-kafkaevent-producer#setup-kafka-receiverapp[Kafka] or https://cloud.ibm.com/docs/codeengine?topic=codeengine-eventing-cosevent-producer#obstorage_ev_app[IBM Cloud Object Storage] and the application by using CloudEvents. These events are sent to a subscribed application as HTTP POST requests. Because the requests are in HTTP format, a RESTful application can receive these events without needing additional libraries and configuration.
```

> To include the CloudEvents restful-ws library within the application

Reviewer: ```suggestion
To include the CloudEvents `restful-ws` library within the application, Add the following https://openliberty.io/guides/maven-intro.html[Maven] dependency to your `pom.xml`:
```

> Add the following https://openliberty.io/guides/maven-intro.html[Maven] dependency to your `pom.xml`

Reviewer: ```suggestion
```

```
<dependency>
    <groupId>io.cloudevents</groupId>
    <artifactId>cloudevents-http-restful-ws-jakarta</artifactId>
    <version>2.5.0</version>
</dependency>
```

Or if you use https://openliberty.io/guides/gradle-intro.html[Gradle]:
```
dependencies {
	implementation group: 'io.cloudevents', name: 'cloudevents-http-restful-ws', version: '2.5.0'}

```

> There are two restful-ws packages provided for CloudEvents

Reviewer: ```suggestion
Two `restful-ws` packages are provided for CloudEvents:
```

* `restful-ws` supports Jakarta EE 8 and the `javax.*` namespace
* `restful-ws-jakarta` supports Jakarta EE 9+ and the `jakarta.*` namespace*

> *Currently there is no 2.5.0 release available, as it is not quite ready for release. It will be the first release of `restful-ws-jakarta` package. If you want to use the new package prior to its release. You can build the package locally and store in a suitable Maven repository for use with your application build. If you build from source, the version will need to be changed to `2.5.0-SNAPSHOT`. The source can be found at https://github.com/cloudevents/sdk-java/tree/master/http/restful-ws-jakarta.

Reviewer: ```suggestion
The CloudEvent 2.5.0 release, the first release of the `restful-ws-jakarta` package, is at the time of this writing  not  yet available. If you want to use the new package before its release, you can build the package locally and store it in a suitable Maven repository for use with your application build. If you build from source,  change the version  to `2.5.0-SNAPSHOT`. You can find the source at https://github.com/cloudevents/sdk-java/tree/master/http/restful-ws-jakarta[HTTP Protocol Binding for Jakarta EE9+ - Jakarta RESTful Web Services].
```

Reviewer: ^^ this part is confusing- the maven/gradle examples show version 2.5.0- but then it says its not currently available and they need to get the source from Git and store it in Maven themselves?

When run within the context of Open Liberty, the `CloudEventsProvider` class is automatically configured for marshalling and unmarshalling of CloudEvents.

> With the libary imported the application can now be updated to make use of CloudEvents.

Reviewer: ```suggestion
With the library imported, you can now update the application to use CloudEvents.
```

> You can view the <<completed_class>> to check that changes made through the following sections align with expected result.

Reviewer: ```suggestion
The following sections guide you through updating the `SystemResource` class from the Getting Started guide  to use CloudEvents. You can view the <<completed_class>> to check that changes you make align with the expected result.
```

==== Returning a CloudEvent

> Having added the necessary library, how do we firstly return a CloudEvent.

Reviewer: ```suggestion
```

> If the Getting Started guide was used, the SystemResource class is a good candidate to convert the response Type from `Responce` to `CloudEvent`

Reviewer: ```suggestion
First, update the `SystemResource` class  response Type from `Response` to `CloudEvent`. In the method declaration, replace `Response` with `CloudEvent`:
```

For the Method Declaration replace `Response` with `CloudEvent`
```
public CloudEvent getProperties() {
```

> Having changed the output type we now need to construct a CloudEvent to return. However we need to do some work on the system properties to be able to include them as the data within the event.

Reviewer: ```suggestion
Now need to construct a CloudEvent to return. However, first we need to do some work on the system properties to be able to include them as the data within the event.
```

> CloudEvents requires the data to be in a binary format when it is provided during its building process as it itself cannot do the conversion from Object to byte[]. As such we can take the properties from `System.getProperties()` and make them into a JSON string with use of Jsonb.

Reviewer: ```suggestion
CloudEvents cannot convert the data from Object to `byte[]` and requires the data to be in a binary format when it is provided during its building process. As such, we can take the properties from `System.getProperties()` and make them into a JSON string by using `Jsonb`.
```

```
/* java.util.properties does not have a direct way to obtain a byte[] so store in an intermediary Map first*/
    Map properties = System.getProperties();
    Jsonb jsonb = JsonbBuilder.create();
/* convert properties map into a JSON string which can then be converted into a byte[]*/
    String jsonString = jsonb.toJson(properties);

```

> With our Data in String format, we can now get the byte[] representation of the data.

Reviewer: ```suggestion
With our data in string format, we can now get the `byte[]` representation of the data.
```

> The `CloudEventBuilder` class provides the necessary components to build our CloudEvent and for this, it is recommended you use `v1()` as that is the most recent specification version.

Reviewer: ```suggestion
The `CloudEventBuilder` class provides the necessary components to build our CloudEvent.  Use the most recent specification version, which is `v1()`.
```

```
    return CloudEventBuilder.v1()
        .withData(jsonString.getBytes())
        .withDataContentType("application/json")
        .withId("properties")
        .withType("java.properties")
        .withSource(URI.create("http://system.poperties"))
        .build();
```

> Aside from `withData()`, the reset of the methods set the values that will be returned as headers in the response.

Reviewer: ```suggestion
Besides `withData()`, the rest of the methods set the values that will be returned as headers in the response.
```

> Once all required properties are set you can build the CloudEvent Object.

Reviewer: ```suggestion
Once all the required properties are set, you can build the CloudEvent Object.
```

> If you are missing any required properties, when your application attempts to build the CloudEvent an exception will be thrown informing of a missing property. To see which properties are required you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification]

Reviewer: ```suggestion
If you are missing any required properties when your application attempts to build the CloudEvent, an exception will be thrown that informs you of a missing property. To see which properties are required, you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification].
```

> As CloudEvents can be from a wide variety of sources can that differ even within a single provider the majority of fields are fairly free-form.

Reviewer: ```suggestion
Because CloudEvents can be from a wide variety of sources that might differ even within a single provider, the majority of the fields are fairly free-form.
```

==== Receiving a CloudEvent

> Having returned a CloudEvent, how can one be handled within the application.

Reviewer: ```suggestion
Now that we've returned a CloudEvent, how is it handled within the application?
```

> For this we are going to enhance the `SystemResource` class to add a query method where we can send a POST request where the body contains the system properties that we want returned.

Reviewer: ```suggestion
First, we'll enhance the `SystemResource` class to add a query method that can send a POST request with a body that contains the system properties that we want returned.
```

> The body we are going to send will be a JSON array with each property we want as an a member of the array

Reviewer: ```suggestion
The body of the request will be a JSON array that contains each property we want as an a member.
```

```
["java.vendor.url","awt.toolkit"]
```

> In the class add the following Method declaration.

Reviewer: ```suggestion
Add the following method declaration to the `SystemResource` class.
```

```

```
    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Timed(name = "queryPropertiesTime",
            description = "Time needed to query the JVM system properties")
    @Counted(absolute = true, description
            = "Number of times the JVM system properties are queried")
    public CloudEvent queryProperties(CloudEvent query){

}
```

> In this case we will be returning a CloudEvent, but you can return any type that you want that confirms the request was received such as `Response.ok().build();`

Reviewer: ```suggestion
In this case, we will return a CloudEvent, but you can return any type that confirms the request was received, such as `Response.ok().build();`
```

> Inside the method, we need to perform some similar actions as we did within `getProperties()`, but also handle the CloudEvent input.

Reviewer: ```suggestion
Inside the method, we need to do some of the same things that we did within  the `getProperties()` method. But we also must handle the CloudEvent input.
```

> As some initial structure add this block to the top of the method

Reviewer: ```suggestion
For some initial structure, add this block to the top of the method.
```

```
Map properties = System.getProperties();
HashMap<String,String> props = new HashMap<>((Map<String,String>)properties);
HashMap<String,String> qProps = new HashMap<String,String>();
Jsonb jsonb = JsonbBuilder.create();
```
This gives us the Map of the properties, but in a form that is more useful later as we need to do more processing then we did within `getProperties()`

> To retrieve the data from the CloudEvent we use `.withData();` to extra and Object of type CloudEventData, the data portion is in binary format, so needs to be converted to make it usable.

Reviewer: ```suggestion
To retrieve the data from the CloudEvent, we use `.withData();` to extra and Object of type CloudEventData, the data portion is in binary format, so needs to be converted to make it usable.
```

Not sure what to suggest here because I don't understand the intended meaning

```
/* Pull data from the data portion of the CloudEvent - this is in binary format so convert it into a standard String*/
CloudEventData data = query.getData();
String jsonString = new String(data.toBytes(), StandardCharsets.UTF_8);
```

> The conversion to a String is the quickest way if you know the input will be in format of JSON, you can check what the sender claims it has sent you can validate via inspect the Data Content Type from `getDataContentType()`

Reviewer: ```suggestion
The conversion to a string is the quickest way if you know the input will be in JSON format. You can check what the sender claims it has sent you can validate via inspect the Data Content Type from `getDataContentType()`
```

again, not sure what to suggest but it looks like several sentences are running together here.

> With the data now in a more usable format we can start to process it and make use of its contents.

Reviewer: ```suggestion
With the data now in a more usable format, we can start to process it and make use of its contents.
```

> As we have a JSON array, we can use jsonb to convert the json to an ArrayList of the keys requested from the properties HashMap.

Reviewer: ```suggestion
Because we have a JSON array, we can use `jsonb` to convert the JSON to an `ArrayList` of the keys that are requested from the properties `HashMap`.
```

```
/* Take the Json Array data and use that to pull out the request properties */
ArrayList<String> tProps = jsonb.fromJson(jsonString, ArrayList.class);
for(String key: tProps){
    qProps.put(key, props.get(key));
}
```
We use the other hashmap created at the start to store the properties we queried for.

> Having built our map of queried properties, it can be returned to the user in the same way we returend

Reviewer: ```suggestion
Now that we have built our map of queried properties, it can be returned to the user in the same way we returned
```

Returned what?

```
return CloudEventBuilder.v1()
    .withData(jsonb.toJson(qProps).getBytes())
    .withDataContentType("application/json")
    .withId("properties")
    .withType("java.properties")
    .withSource(URI.create("http://system.poperties"))
    .build();
```

> With that the new method is ready to be tested.

Reviewer: ```suggestion
Next, we're ready to test the new method.
```

> To invoke the new method, we make a POST request against `/dev/system/properties` with the HTTP request being a CloudEvent. This can be done locally, or against a rebuild Docker image.

Reviewer: ```suggestion
To invoke the new method, we make a POST request against `/dev/system/properties` with the HTTP request being a CloudEvent. You can do this locally, or against a rebuilt Docker image.
```

> To invoke this method use the following curl command:

Reviewer: ```suggestion
To invoke this method, use the following curl command:
```

```
curl -X POST http://${url}/dev/system/properties \
-H "Ce-Specversion: 1.0" \
-H "Ce-Type: properties" \
-H "Ce-Source: io.cloudevents.examples/properties" \
-H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f78" \
-H "Content-Type: application/json" \
-H "Ce-Subject: resources" \
-d "[\"java.vendor.url\",\"awt.toolkit\"]"
```

> In the same way when we returned a CloudEvent, when we make the request, we need to provide the required set of headers so that the application can correctly convert the request into a CloudEvent.

Reviewer: does this refer to the previous snippet?

        HashMap<String,String> props = new HashMap<>((Map<String,String>)properties);
        HashMap<String,String> qProps = new HashMap<String,String>();
        Jsonb jsonb = JsonbBuilder.create();

        /* Pull data from the data portion of the CloudEvent - this is in binary format so convert it into a standard String*/
        CloudEventData data = query.getData();
        String jsonString = new String(data.toBytes(), StandardCharsets.UTF_8);

        /* Take the Json Array data and use that to pull out the request properties */
        ArrayList<String> tProps = jsonb.fromJson(jsonString, ArrayList.class);
        for(String key: tProps){
            qProps.put(key, props.get(key));
        }

        /* return a CloudEvent with our queried properties */
        return CloudEventBuilder.v1()
                .withData(jsonb.toJson(qProps).getBytes())
                .withDataContentType("application/json")
                .withId("properties")
                .withType("java.properties")
                .withSource(URI.create("http://system.poperties"))
                .build();
    }

}
```

=== Updating your IBM Cloud Code Engine application

> With the application now capable of returning and receiving CloudEvents we can upade our application in ICCE

Reviewer: ```suggestion
The application can now return and receive CloudEvents.  We're can update our application in ICCE.
```

To update your application

> * Rebuild your Liberty application with the CloudEvent changes

Reviewer: ```suggestion
1. Rebuild your Liberty application with the CloudEvent changes.
```

> * Rebuild your docker container and publish to ICCR either updating the version or leave as is if you are using the image hash

Reviewer: ```suggestion
2. Rebuild your docker container and publish to it ICCR, either by updating the version or leaving it as is if you are using the image hash
```

> * https://cloud.ibm.com/docs/codeengine?topic=codeengine-update-app#update-app-crimage-console[Update your ICCE application] to use the new version

Reviewer: ```suggestion
3. https://cloud.ibm.com/docs/codeengine?topic=codeengine-update-app#update-app-crimage-console[Update your ICCE application] to use the new version.
```

Reviewer: new version of what?

> With your application updated you can validate your changes in ICCE by invoking using the previous curl commands.

Reviewer: which curl commands?
